flang/lib/Lower/Bridge.cpp

```diff
@@ -395,7 +396,20 @@ class FirConverter : public Fortran::lower::AbstractConverter {
                                fir::getTypeParams(hexv));
           return fir::substBase(hexv, temp);
         });
-    return bindSymbol(sym, exval);
+
+    bool bindStatus = bindSymbol(sym, exv);
+    if (bindStatus && init) {
+      if (auto seqTy = symType.dyn_cast<fir::SequenceType>()) {
```

Comment:
Is this check sufficient (array vs not array)? Or do we have to do a case analysis based handling for the various kinds of ExtendedValue?

Reply:
The allocatable case that is probably not covered correctly with this approach (it will fall in the else). I am not sure what are OpenMP spec here. I would assume the private copy will need to be allocated if the outer variable is allocated and the value must be then be copied.
So you may want to go through the ExtendedValue to handle the allocatable case in a custom way (fir::MutableBoxValue case, you can make the difference with pointer with `fir::MutableBoxValue::isAllocatable()`).

Reply:
Thanks @jeanPerier. I will try this.

Reply:
Is f2003 feature, reallocate on assign already implemented?

Reply:
A simple load & store won't be sufficient for derived types after F95. It could also be problematic for CHARACTER.

Reply:
> Is f2003 feature, reallocate on assign already implemented ?

Yes, it is implemented [here for scalar allocatables](https://github.com/flang-compiler/f18-llvm-project/blob/79dd7cc4127fb4dffea14350a21c7833451c1248/flang/lib/Lower/Bridge.cpp#L1842), and [here for array allocatables](https://github.com/flang-compiler/f18-llvm-project/blob/79dd7cc4127fb4dffea14350a21c7833451c1248/flang/lib/Lower/ConvertExpr.cpp#L2556). The common core of the implementation is `Fortran::lower::genReallocIfNeeded` declared [here](https://github.com/flang-compiler/f18-llvm-project/blob/79dd7cc4127fb4dffea14350a21c7833451c1248/flang/include/flang/Lower/Allocatable.h#L139).

Reply:
OK.
I have just made an update to the patch so that the code generated will correctly load from an allocatable. I will fix the allocation and re-allocation in a separate patch.

Reply:
I have removed the handling of allocatables from this patch.